File: backend/python/common/parent_watch_test.py
"""Unit tests for the parent-death watcher (parent_watch.py).

Run standalone (Python standard library only, no backend venv needed):
    python3 -m unittest parent_watch_test

The core test (test_detects_reparent) builds a genuine two-level process tree
(test -> middle -> grandchild) with os.fork, lets the middle process die, and
asserts the grandchild's parent_watch._watch detects the reparenting and
self-terminates — mirroring the Go test in pkg/grpc/parentwatch_test.go and the
C++ test in backend/cpp/llama-cpp/parent_watch_test.cpp.
"""

import os
import sys
import tempfile
import threading
import time
import unittest

import parent_watch


class TestParentWatchEnvParsing(unittest.TestCase):
    def setUp(self):
        self._saved = {
            k: os.environ.get(k)
            for k in (parent_watch.ENV_PARENT_WATCH, parent_watch.ENV_PARENT_WATCH_INTERVAL)
        }
        for k in self._saved:
            os.environ.pop(k, None)

    def tearDown(self):
        for k, v in self._saved.items():
            if v is None:
                os.environ.pop(k, None)
            else:
                os.environ[k] = v

    def test_interval_default(self):
        self.assertEqual(parent_watch._interval_seconds(), 2.0)

    def test_interval_units(self):
        cases = {"500ms": 0.5, "2s": 2.0, "1m": 60.0, "3": 3.0, "0.5s": 0.5}
        for raw, expected in cases.items():
            os.environ[parent_watch.ENV_PARENT_WATCH_INTERVAL] = raw
            self.assertAlmostEqual(parent_watch._interval_seconds(), expected, msg=raw)

    def test_interval_garbage_falls_back(self):
        os.environ[parent_watch.ENV_PARENT_WATCH_INTERVAL] = "garbage"
        self.assertEqual(parent_watch._interval_seconds(), 2.0)

    @unittest.skipIf(os.name == "nt" or sys.platform.startswith("win"), "POSIX only")
    def test_enabled_default(self):
        self.assertTrue(parent_watch._enabled())

    @unittest.skipIf(os.name == "nt" or sys.platform.startswith("win"), "POSIX only")
    def test_disabled_by_falsey(self):
        for val in ("false", "0", "no", "off", "OFF", " False "):
            os.environ[parent_watch.ENV_PARENT_WATCH] = val
            self.assertFalse(parent_watch._enabled(), msg=val)

    @unittest.skipIf(os.name == "nt" or sys.platform.startswith("win"), "POSIX only")
    def test_enabled_by_truthy(self):
        for val in ("true", "1", "yes", "on"):
            os.environ[parent_watch.ENV_PARENT_WATCH] = val
            self.assertTrue(parent_watch._enabled(), msg=val)


@unittest.skipIf(os.name == "nt" or sys.platform.startswith("win"), "fork/reparent is POSIX only")
class TestParentWatchReparent(unittest.TestCase):
    def _wait_for_file(self, path, timeout=10.0):
        deadline = time.time() + timeout
        while time.time() < deadline:
            if os.path.exists(path):
                return True
            time.sleep(0.02)
        return False

    def test_detects_reparent(self):
        tmpdir = tempfile.mkdtemp(prefix="parentwatch_test_")
        ready_file = os.path.join(tmpdir, "ready")
        exited_file = os.path.join(tmpdir, "exited")

        middle = os.fork()
        if middle == 0:
            # ---- middle process ----
            grandchild = os.fork()
            if grandchild == 0:
                # ---- grandchild process: arm the REAL watcher against middle ----
                orig_ppid = os.getppid()

                def on_death():
                    with open(exited_file, "w") as f:
                        f.write("1")
                    os._exit(7)

                threading.Thread(
                    target=parent_watch._watch,
                    args=(orig_ppid, 0.05, on_death),
                    daemon=True,
                ).start()

                # Safety valve: never linger if something goes wrong.
                def bail():
                    time.sleep(30)
                    os._exit(2)

                threading.Thread(target=bail, daemon=True).start()

                # Signal readiness only after the watcher captured orig_ppid.
                with open(ready_file, "w") as f:
                    f.write(str(os.getpid()))
                while True:
                    time.sleep(1)
            else:
                # middle: wait until grandchild is ready, then exit to orphan it.
                if not self._wait_for_file(ready_file):
                    os._exit(5)
                os._exit(0)

        # ---- test (top) process ----
        os.waitpid(middle, 0)  # reap middle only; grandchild is orphaned

        self.assertTrue(os.path.exists(ready_file), "grandchild never signaled readiness")
        self.assertTrue(
            self._wait_for_file(exited_file),
            "watcher did not detect parent death within timeout",
        )

        # Best-effort cleanup: kill the grandchild if it somehow survived.
        try:
            with open(ready_file) as f:
                pid = int(f.read().strip())
            if pid > 1:
                os.kill(pid, 9)
        except (OSError, ValueError):
            pass
        for p in (ready_file, exited_file):
            try:
                os.remove(p)
            except OSError:
                pass
        try:
            os.rmdir(tmpdir)
        except OSError:
            pass


if __name__ == "__main__":
    unittest.main()

File: backend/cpp/llama-cpp/parent_watch_test.cpp
// Unit tests for the parent-death watcher (parent_watch.h).
//
// Build & run standalone (C++ standard library only, no nlohmann/json needed):
//   g++ -std=c++17 -pthread parent_watch_test.cpp -o t && ./t
//
// The core test (TestDetectsReparent) builds a genuine two-level process tree
// (test -> middle -> grandchild), lets the middle process die, and asserts the
// grandchild's watch_parent_death detects the reparenting and self-terminates —
// mirroring the Go test in pkg/grpc/parentwatch_test.go, but with fork(2).
//
// On Windows this file compiles to a no-op success (the watcher is unsupported
// there), matching parent_watch.h's platform gating.

#include <cstdio>
#include <cstdlib>
#include <string>

#include "parent_watch.h"

static int failures = 0;

static void check(bool ok, const std::string &name) {
    if (!ok) {
        failures++;
        fprintf(stderr, "FAIL: %s\n", name.c_str());
    } else {
        fprintf(stderr, "ok:   %s\n", name.c_str());
    }
}

// Env-parsing tests are platform-independent and always run.
static void test_env_parsing() {
    using namespace llama_grpc;

    // Interval: default when unset.
    unsetenv("LOCALAI_BACKEND_PARENT_WATCH_INTERVAL");
    check(parent_watch_interval_ms() == 2000, "interval default 2000ms");

    setenv("LOCALAI_BACKEND_PARENT_WATCH_INTERVAL", "500ms", 1);
    check(parent_watch_interval_ms() == 500, "interval 500ms");

    setenv("LOCALAI_BACKEND_PARENT_WATCH_INTERVAL", "2s", 1);
    check(parent_watch_interval_ms() == 2000, "interval 2s");

    setenv("LOCALAI_BACKEND_PARENT_WATCH_INTERVAL", "1m", 1);
    check(parent_watch_interval_ms() == 60000, "interval 1m");

    setenv("LOCALAI_BACKEND_PARENT_WATCH_INTERVAL", "3", 1); // bare number -> seconds
    check(parent_watch_interval_ms() == 3000, "interval bare 3 -> 3000ms");

    setenv("LOCALAI_BACKEND_PARENT_WATCH_INTERVAL", "garbage", 1);
    check(parent_watch_interval_ms() == 2000, "interval garbage -> default");
    unsetenv("LOCALAI_BACKEND_PARENT_WATCH_INTERVAL");

#if !defined(_WIN32)
    // Enabled semantics (POSIX only; always false on Windows).
    unsetenv("LOCALAI_BACKEND_PARENT_WATCH");
    check(parent_watch_enabled(), "enabled by default");

    for (const char *falsey : {"false", "0", "no", "off", "OFF", " False "}) {
        setenv("LOCALAI_BACKEND_PARENT_WATCH", falsey, 1);
        check(!parent_watch_enabled(), std::string("disabled by '") + falsey + "'");
    }
    setenv("LOCALAI_BACKEND_PARENT_WATCH", "true", 1);
    check(parent_watch_enabled(), "enabled by 'true'");
    setenv("LOCALAI_BACKEND_PARENT_WATCH", "1", 1);
    check(parent_watch_enabled(), "enabled by '1'");
    unsetenv("LOCALAI_BACKEND_PARENT_WATCH");
#endif
}

#if !defined(_WIN32)

#include <atomic>
#include <ctime>
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>

static bool file_exists(const std::string &p) {
    struct stat st;
    return ::stat(p.c_str(), &st) == 0;
}

static bool wait_for_file(const std::string &p, int timeout_ms) {
    int waited = 0;
    while (waited < timeout_ms) {
        if (file_exists(p)) {
            return true;
        }
        usleep(20 * 1000);
        waited += 20;
    }
    return false;
}

static void write_file(const std::string &p, const std::string &content) {
    FILE *f = fopen(p.c_str(), "w");
    if (f) {
        fwrite(content.data(), 1, content.size(), f);
        fclose(f);
    }
}

// Builds test -> middle -> grandchild via fork(2). The grandchild arms the REAL
// watch_parent_death against middle; middle exits, orphaning the grandchild;
// the watcher must detect the reparenting and self-terminate.
static void test_detects_reparent() {
    char tmpl[] = "/tmp/parentwatch_test_XXXXXX";
    char *dir = mkdtemp(tmpl);
    if (dir == nullptr) {
        check(false, "mkdtemp");
        return;
    }
    const std::string ready_file = std::string(dir) + "/ready";
    const std::string exited_file = std::string(dir) + "/exited";

    pid_t middle = fork();
    if (middle < 0) {
        check(false, "fork middle");
        return;
    }

    if (middle == 0) {
        // ---- middle process ----
        pid_t grandchild = fork();
        if (grandchild < 0) {
            _exit(4);
        }
        if (grandchild == 0) {
            // ---- grandchild process ----
            pid_t orig_ppid = getppid(); // == middle
            std::thread([&]() {
                llama_grpc::watch_parent_death(orig_ppid, 50 /*ms*/, [&]() {
                    write_file(exited_file, "1");
                    _exit(7);
                });
            }).detach();

            // Safety valve: never linger if something goes wrong.
            std::thread([]() {
                usleep(30 * 1000 * 1000);
                _exit(2);
            }).detach();

            // Signal readiness only after the watcher captured orig_ppid.
            write_file(ready_file, std::to_string(getpid()));
            for (;;) {
                pause();
            }
        }
        // middle: wait until grandchild is ready, then exit to orphan it.
        if (!wait_for_file(ready_file, 10000)) {
            _exit(5);
        }
        _exit(0);
    }

    // ---- test (top) process ----
    int status = 0;
    waitpid(middle, &status, 0); // reap middle only; grandchild is orphaned

    check(file_exists(ready_file), "grandchild signaled readiness");

    bool detected = wait_for_file(exited_file, 10000);
    check(detected, "watcher detected parent death and self-terminated");

    // Best-effort cleanup: kill the grandchild if it somehow survived.
    if (file_exists(ready_file)) {
        FILE *f = fopen(ready_file.c_str(), "r");
        if (f) {
            int pid = 0;
            if (fscanf(f, "%d", &pid) == 1 && pid > 1) {
                kill(pid, SIGKILL);
            }
            fclose(f);
        }
    }
    unlink(ready_file.c_str());
    unlink(exited_file.c_str());
    rmdir(dir);
}

#endif // !_WIN32

int main() {
    test_env_parsing();
#if !defined(_WIN32)
    test_detects_reparent();
#endif
    if (failures == 0) {
        fprintf(stderr, "\nAll parent_watch tests passed.\n");
        return 0;
    }
    fprintf(stderr, "\n%d parent_watch test(s) failed.\n", failures);
    return 1;
}
